etc/i2c/eeprom.py

```diff
@@ -18,18 +19,24 @@ def read_bytes(bus, device, length = 32):
 # access methods.  In particular, the 24c64 needs a two byte read address to be
 # written to it, whereas the 24c02 interprets a two byte write as writing a byte
 # to the EEPROM.
-#    The code below relies on undocumented behaviour that we are seeing on the
-# M24C64: writing a single byte appears to reset the entire page address (32 bit
-# transfer pages), but not the byte offset within the page.
-#    Therefore the trick here is to read half of the same page twice using
-# single byte addressing: if we get the same result both times then assume that
-# this is ok, otherwise fall back to two byte addressing.
+# Here we attempt to read the IPMI header twice in 8-bit mode.
+# If we get the header successfully twice then 8-bit addressing worked.
+# If not we assume the device is 16-bit addressed 
+# NOTE: this fails for an 8-bit device which doesn't contain valid data
+
 def detect_16bit(device = 0x50):
-    # Read two half pages.  If they are the same then we can assume that we were
-    # able to reset
-    first_read  = read_8bit_address(device, length = 15)
-    second_read = read_8bit_address(device, length = 15)
-    return first_read != second_read
+    first_read  = read_8bit_address(device, length = 8)
+    second_read = read_8bit_address(device, length = 8)
```

Comment:
I might be missing something here, but doesn't this require the address counter to be at the start like the old code did?

Reply:
read_8bit_address attempts to read data starting from zero (sends message to set address to zero). We attempt to read sending an 8bit address twice (if the counter was at zero, the first read would return the header but the second wouldn't), and if we don't get the header data twice we assume it didn't accept the address send and hence is 16bit. We then move on to check with a 16-bit read, and error if we still don't get a valid header.

Reply:
I see, that makes sense

Reply:
Could you put this text in as a comment please?